Implement IMT beamforming normalization

CalilQ opened the pull request:
The script in `normalize_script.py` calls the class `BeamformingNormalizer`, which calculates the correction factors and saves them in a `.npz` file.
If parameter `beamforming_normalization` is set to `true`, the simulator loads the data in `bs_normalization_file` and `ue_normalization_file` and uses the correction factors in `AntennaBeamformingImt` class gain calculation.

This branch can be deleted after the merge.

CalilQ committed:
Implement calculate_gain_matrix method

CalilQ committed:
Implement calculate_correction_factor

Also remove calculate_gain_matrix

CalilQ committed:
Implement generate_correction_matrix

CalilQ committed:
Implement _save_files method

CalilQ committed:
Implement _save_files method

CalilQ committed:
Fix antenna/antenna_beamforming_imt.py line skip

CalilQ committed:
Merge branch 'dev-beamforming-norm' of https://github.com/SIMULATOR-WG/SHARC into dev-beamforming-norm

CalilQ committed:
BeamformingNormalization test set up and test construction

Minor construction bug also fixed, as well as package created

CalilQ committed:
Specify AntennaElementImtM2101 gain unit of measure

It's dBi

CalilQ committed:
Write calculate_correction_factor initial tests

Result still different from expected value though

CalilQ committed:
Fix correction factor integral

Calculation was mixing degrees and radians

CalilQ committed:
Implement initial version of second normalization test

CalilQ committed:
Perform integration based on relative error

CalilQ committed:
Merge branch 'development' of https://github.com/SIMULATOR-WG/SHARC into dev-beamforming-norm

CalilQ committed:
Reduce integral epsilon for convergence

Integrals do not converge when tolerance is comparable to integrand value

CalilQ committed:
Fix correction factor sign

Correction factor is the inverse of the total integrated gain

CalilQ committed:
Create tolerance attribute and test generate_correction_matrix

Minor bug fixes as well

CalilQ committed:
Improve test_generate_correction_matrix

Add another test to increase coverage

CalilQ committed:
Change class name and implement main

CalilQ committed:
Improve BeamformingNormalizer plot

CalilQ committed:
Document class

CalilQ committed:
Save grid parameters as well

CalilQ committed:
Create normalization script

Normalization matrix files are saved with a hash of its parameters as name.

CalilQ committed:
Change resolution and tolerance

CalilQ committed:
Pass normalization parameter to AntennaBeamformingImt

CalilQ committed:
Account for normalization in script

CalilQ committed:
Implement correction factor usage

CalilQ committed:
Fix unit tests

All test now pass

CalilQ committed:
Delete test npz file

CalilQ committed:
Fix normalize script AntennaPar

They were considering the normalization in the calculation.

CalilQ committed:
Improve normalization file path

CalilQ committed:
Create normalization tests

CalilQ committed:
Save both normalization factors in the same file

CalilQ committed:
Implement correction factor file parameters

CalilQ committed:
Adapt normalize_script.py to new file save structure

CalilQ committed:
Fix unit tests

Tests need to account for normalization file

CalilQ committed:
Remove unnecessary imports

CalilQ committed:
Fix normalization file name

'sharc' folder name is not necessary

CalilQ committed:
Improve AntennaBeamformingImt construction efficiency

CalilQ committed:
Change normalization data to a dict

CalilQ committed:
Use stored data in calculate_gain

CalilQ committed:
Fix test_antenna_beamforming_imt.py to account for dict normalization data

CalilQ committed:
Adapt normalize_script.py to dict normalization data

anollba reviewed (commented):
usar nomes de variáveis mais fáceis de entender (o que é res? o que é cf?)

anollba reviewed (commented):
dummy

anollba reviewed (commented):
incluir alguns comentários, em particular sobre a necessidade de se gerar os arquivos externos antes, e como gerá-los

anollba reviewed (commented):
what's an "escan value" in BeamformingAnalyser. Is it azimuth angle?

CalilQ committed:
Change variables names

cf was changed to correction_factor, res was changed to resolution and tol was changed to tolerance

CalilQ committed:
Improve beamforming normalization variable names and documentation

CalilQ commented:
Variable names and comments improved in commits 533a6d709337184f16ef994bc033ca88206f3e43 and 4592d807ad65e530e637cf1e4860bb3dd3669342
